### PR TITLE
Negated ceg nodes dashed

### PR DESCRIPTION
**Pull request by sebeder**
This introduces a popup menu allowing for negating CEG connections.

### CODE REVIEW

File: web/src/app/modules/views/main/editors/modules/graphical-editor/providers/properties/shape-provider.ts [hunk 1]
```diff
@@ -36,7 +39,7 @@ export class ShapeProvider extends ProviderBase {
 
         this.shapeMap[ProcessStart.className] = {
             // tslint:disable-next-line:max-line-length
```
Comment (wehrlean) [inline web/src/app/modules/views/main/editors/modules/graphical-editor/providers/properties/shape-provider.ts]: Remove linter-comment

File: web/src/app/modules/views/main/editors/modules/graphical-editor/providers/properties/shape-provider.ts [hunk 2]
```diff
@@ -48,7 +51,7 @@ export class ShapeProvider extends ProviderBase {
 
         this.shapeMap[ProcessEnd.className] = {
             // tslint:disable-next-line:max-line-length
```
Comment (wehrlean) [inline web/src/app/modules/views/main/editors/modules/graphical-editor/providers/properties/shape-provider.ts]: Remove linter-comment

File: web/src/app/modules/views/main/editors/modules/graphical-editor/components/util/change-translator.ts [hunk 3]
```diff
@@ -18,6 +18,8 @@ import { ValuePair } from '../../providers/properties/value-pair';
 import { Type } from 'src/app/util/type';
 import { CEGNode } from 'src/app/model/CEGNode';
 import { type } from 'os';
```
Comment (wehrlean) [inline web/src/app/modules/views/main/editors/modules/graphical-editor/components/util/change-translator.ts]: Remove unused import

File: web/src/app/modules/views/main/editors/modules/graphical-editor/components/util/change-translator.ts [hunk 4]
```diff
@@ -61,9 +63,35 @@ export class ChangeTranslator {
...
+            let changeMade = false;
+            if (Arrays.contains(newStyles, EditorStyle.ADDITIONAL_CEG_CONNECTION_NEGATED_STYLE)) {
+                connection.negate = true;
+            } else if (Arrays.contains(removedStyles, EditorStyle.ADDITIONAL_CEG_CONNECTION_NEGATED_STYLE)) {
+                connection.negate = false;
+            }
+
+            if (changeMade) {
+                await this.dataService.updateElement(connection, true, Id.uuid);
+            }
```
Comment (wehrlean) [inline web/src/app/modules/views/main/editors/modules/graphical-editor/components/util/change-translator.ts]: Isn't _changeMade_ always false, thus the _.updateElement_ is never called?

File: web/src/app/modules/views/main/editors/modules/graphical-editor/components/graphical-editor.component.ts [hunk 5]
```diff
@@ -323,23 +336,19 @@ export class GraphicalEditor {
       this.graph.setCellWarning(vertex, null);
     }
     const validationResult = this.validationService.getValidationResults(this.model);
-    const invalidNodes = validationResult.filter(e => e.element.className === 'ProcessDecision'
-      || e.element.className === 'ProcessStart'
-      || e.element.className === 'ProcessEnd'
-      || e.element.className === 'ProcessStep'
-      || e.element.className === 'CEGNode');
+    const invalidNodes = validationResult.filter(e => this.elementProvider.isNode(e.element));
```
Comment (wehrlean) [inline web/src/app/modules/views/main/editors/modules/graphical-editor/components/graphical-editor.component.ts:479]: Good to know, thanks :)

File: web/src/app/modules/views/main/editors/modules/graphical-editor/components/editor-components/editor-popup.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        this.graph.popupMenuHandler['factoryMethod'] = this.provideMenu.bind(this);
+    }
+
+    private async provideMenu(menu: mxgraph.mxPopupMenuHandler, cell: mxgraph.mxCell, evt: PointerEvent) {
+
+        if (cell === undefined || cell === null) {
+            return;
+        }
+
+        menu.addItem('Delete', null, () => {
```
Comment (wehrlean) [inline web/src/app/modules/views/main/editors/modules/graphical-editor/components/editor-components/editor-popup.ts]: Add String to i18n

File: web/src/app/modules/views/main/editors/modules/graphical-editor/components/editor-components/editor-popup.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            this.graph.removeCells([cell]);
+        }, undefined, undefined, undefined, undefined);
+
+        const element = await this.dataService.readElement(cell.id, true);
+        if (Type.is(element, CEGConnection)) {
+
+            const connection = element as CEGConnection;
+            const icon = connection.negate ? 'fa fa-check' : 'fa fa-circle-o';
+
+            menu.addItem('Negate', null, async () => {
```
Comment (wehrlean) [inline web/src/app/modules/views/main/editors/modules/graphical-editor/components/editor-components/editor-popup.ts]: Add String to i18n

File: web/src/app/modules/views/main/editors/modules/graphical-editor/components/editor-components/editor-popup.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+                    StyleChanger.removeStyle(cell, this.graph, EditorStyle.ADDITIONAL_CEG_CONNECTION_NEGATED_STYLE);
+                } else {
+                    StyleChanger.addStyle(cell, this.graph, EditorStyle.ADDITIONAL_CEG_CONNECTION_NEGATED_STYLE);
+                }
+                this.graph.getSelectionModel().clear();
+                this.graph.getModel().endUpdate();
+            }, undefined, icon, undefined, undefined);
+        }
+    }
+
```
Comment (wehrlean) [inline web/src/app/modules/views/main/editors/modules/graphical-editor/components/editor-components/editor-popup.ts]: Generally remove useless empty lines

**Review by wehrlean — CHANGES_REQUESTED** [review]
I like the popups, really nice
Some smaller and bigger issues:

- Two-way-binding with the negate-button doesn't work
- Style of the popup doesn't fit to the rest of Specmate, I think
- Clicking left on the label (variable or condition) of a node throws an error
- Add a new connection and then negate it twice leads to an error
- The popup stays on screen, if you open it and then switch the model 
- The other comments

**Comment by sebeder** [comment]
I adressed all issues, but "the popup stays on screen...", because to realize this, we'd need a global reference to the popup and a global listener for click-events. Both is ugly and I'd suggest we live with this small glitch. OK?

**Review by wehrlean — APPROVED** [review]
Nice, everything works. Only one remark. At the beginning, I don't realize, that the circle and the checkmark is a kind of checkbox. Thus the 'negated' is a little bit confusing. Especially the 'delete' is a command and the 'negated' is a status. I expected 'negate' as a command.
And the popup-problem with new screen is no problem. This happens very rarely.

**Pull request merged**